Change the way workflows are stored in the database from XML to tables

Arnei opened the pull request:
This PR is in preperation for:
- #3377!

<hr>

Currently, workflows are stored as a single XML object in the database. This makes it difficult/inperformant to use more complex queries, which is why workflow information is also stored in a Solr Index. In preparation for removing Solr from the Workflow Service, this PR creates new database tables for workflow information.

Key changes:
- Workflow information is now stored in tables named "oc_workflow*"
- Workflow information is no longer stored in the job tables "payload" column. Instead, it stores the workflow_id there (which is the same id as the corresponding "START_WORKFLOW" job, so it is actually kinda superfluous)
- Therefore, the WorkflowInstance and WorkflowOperationInstance classes are now using JPA annotations instead of JAXB annotations.
- Workflows still require XML representations for e.g. REST endpoints, so corresponding JAXB classes for that purposes are added
- Requires a db upgrade script. The script is a python script, placed in the usual sport for db upgrade scripts.
- Adds an endpoint that allows to query for workflow instances by mediapackage id

Work sponsored by our generous Crowdfunding initiatives!

<hr>

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

Arnei committed:
Attempt at replacing the workflow XML in the job table payload column with a new database table for workflow information

Arnei committed:
Attempt at replacing the workflow operation and configuration XML in the workflow table with their own tables

Arnei committed:
Fixed JPA issues that prevented db tables from being created

Arnei committed:
- Fix checkstyle things that broke the code arbitrarily (mostly import ordering)
- Fix dollar sign replacement for execution conditions
- Fix inserting new operations into workflow (position and order)

Arnei committed:
- Temp fix for "$()" strings in config values

Arnei committed:
Fix Workflow Display in Admin UI

Arnei committed:
Introduce Jaxb classes to fix SolrIndex and RestService

- RestService still not completely fixed, some requests fail while a workflow is running

Arnei committed:
Fix compilation errors

Arnei committed:
Fixed Rest Endpoints

- Fixed remove operation not working
- Fixed getting all workflows causing error messages

Arnei committed:
Configuration values can be very long

Arnei committed:
Cleanup

Arnei committed:
Added comments and exceptions to workflow database implementation

Arnei committed:
Turn WorkflowConfiguration.java from useless interface to more neat abstract class

Arnei committed:
Fixed not updating of mp metadata in wf instance

The missing update statement for example caused metadata from a dc catalog to not be transferred into the mediapackage xml of the workflow instance, which could cause errors.

Also fixes a response error where the workflow instance was not properly marshalled.

Arnei committed:
Fixed TODO "Order by date created"

Arnei committed:
Fixed TODO "Should this db query filter by organization"

Arnei committed:
Fixed TODO "Check authorization"

Arnei committed:
Fixed TODO "Store organization in workflow"

Arnei committed:
Rewrote TODO "Make band-aid more elegant"

Arnei committed:
Fixed TODO "workflow operation position fix"

Arnei committed:
Make database table names more concise

Arnei committed:
Removed execution history from workflow operations as it is not used anywhere.

Arnei committed:
Avoid unnecessary iterations when marshalling configurations

- Additional cleanup

Arnei committed:
Add migration script

Arnei committed:
Fix tests in workflow-service-impl

By either rewriting the tests, or fixing actual errors in the code

Arnei committed:
Fix tests in admin-ui

Arnei committed:
Fix tests in workflow-workflowoperation

Arnei committed:
Fix tests in videoeditor

Arnei committed:
Remove debugging log statements

Arnei committed:
Update workflow db migration script to remove duplicate information from oc_jobs

Also adds some error handling and logging

Arnei committed:
Added currently declared indexes to workflow db migration script

Arnei committed:
Does not work because xml marshalling maps is non trivial

Arnei committed:
Reintroduced JaxbWorkflowConfiguration, which makes mapping easier

Way easier way to map between the JPA Map and JAXB Set

Arnei committed:
Fix db upgrade script

Arnei committed:
Add warning to db upgrade script

Arnei committed:
Merge branch 'develop' into workflows_from_xml_to_database_tables

# Conflicts:
#	modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowParser.java

Arnei committed:
Fixed tests

Arnei committed:
Unified database queries that pertain to active workflows for a mediapackage

Arnei committed:
Fix workflow db upgrade script

Arnei committed:
Added postgresql version of the workflow db upgrade script

Arnei committed:
Added release and upgrade notes

Arnei committed:
Allow the value of a workflow configuration key-value-pair to be null

Arnei committed:
Updated notes on API changes

Arnei committed:
Fix trailing whitespaces in docs

Arnei committed:
Merge remote-tracking branch 'upstream/develop' into workflows_from_xml_to_database_tables

# Conflicts:
#	docs/guides/admin/docs/releasenotes.md

Arnei committed:
Merge remote-tracking branch 'upstream/develop' into workflows_from_xml_to_database_tables

# Conflicts:
#	docs/guides/admin/docs/releasenotes.md

Arnei committed:
Missed some files while merging

Arnei committed:
Added a more robust function for the IndexService to get the currently running workflow

Arnei committed:
Removed duplicate WorkflowDatabaseException class

Arnei committed:
Cleanup

Arnei committed:
Merge remote-tracking branch 'upstream/develop' into workflows_from_xml_to_database_tables

# Conflicts:
#	modules/index-service/pom.xml
#	modules/index-service/src/main/java/org/opencastproject/index/service/resources/list/provider/WorkflowsListProvider.java

Arnei committed:
Merge remote-tracking branch 'upstream/develop' into workflows_from_xml_to_database_tables

# Conflicts:
#	modules/index-service/pom.xml

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+
+# Cleanup artifacts from previous runs of this script
+print("Clearing out potential artifacts from previous runs...")
+delete_workflow_table = f"DROP TABLE {workflow_table_name}"
```
We probably shouldn't silently delete database tables. If there is already a table like this, something is wrong and we should fail and have the operator take a look. The same goes for the lines below

lkiesow commented on docs/guides/admin/docs/releasenotes.md:
```diff
@@ -1,131 +1,32 @@
-  avoid performance problems and remove incorrect and/or misleading data.
+- Internally, workflows were stored as huge xml objects in the job table in the databaase. With Opencast 12,
+workflows now get their own tables in the database. This makes it easier and quicker to query for workflows when
+not using the Solr Index.
```
Move to https://github.com/opencast/opencast/tree/develop/docs/guides/admin/docs/releasenotes

lkiesow commented on docs/guides/admin/docs/upgrade.md:
```diff
@@ -10,76 +10,11 @@ please refer to [older release notes](https://docs.opencast.org).
 
-Upgrading to Opencast 11 requires a DB migration, as some tables have changed slightly.
-Migration scripts can be found in `doc/upgrade/10_to_11/`.
+Upgrading to Opencast 12 requires a DB migration, as some tables have changed slightly.
```
Move to https://github.com/opencast/opencast/tree/develop/docs/guides/admin/docs/releasenotes

lkiesow replied:
Also, there are no such SQL scripts. There are Python scripts, however, which users need to modify before they can run then which is not documented.

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+import sys
+import xml.etree.ElementTree as ET
+import distutils.util
+from distutils.util import strtobool
```
```
/home/lars/dev/opencast/docs/upgrade/11_to_12/workflows_mysql5.py:15: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
  import distutils.util
```

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+# Upgrade script for MySQL databases
+# Requires Python3 to run
+# Required packages:
+#   $ pip install mysql-connector-python
```
It would be great for this to appear in the upgrade guide but also if this folder would have a `readme.md` file mentioning that users need to edit this file and install dependencies.

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+# Upgrade script for MySQL databases
```
While it is basically the same, since we officially dropped MySQL support but kept MariaDB support, maybe switch from MySQL to MariaDB in naming?

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+password = "dbpassword"
+host = "127.0.0.1"
+database = "opencast"
+
```
```suggestion

# DO NOT EDIT ANYTHING BELOW

```

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+from mysql.connector import Error
+import sys
+import xml.etree.ElementTree as ET
+import distutils.util
```
```suggestion
```

```
❯ flake8 workflows_mysql5.py                                                                                                                                                                                                                  
workflows_mysql5.py:12:1: F401 'sys' imported but unused                                                                                                                                                                                      
workflows_mysql5.py:14:1: F401 'distutils.util' imported but unused  
```

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+# Module Imports
+import mysql.connector
+from mysql.connector import Error
+import sys
```
```suggestion
```

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+        )
+        connection.row_factory = lambda cursor, row: row[0]
+        print("Connection to MySQL DB successful")
+    except Error as e:
+        print(f"The error '{e}' occurred")
+    return connection
```
Why catch this and return `None`? Shouldn't the script immediately fail if a connection could not be established? I suggest removing the `try … except …` More so since you just catch the error to then print the error anyway. Something which would happen automatically.

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+def execute_query(connection, query):
+    cursor = connection.cursor()
+
+    try:
+        cursor.execute(query)
+        connection.commit()
+        print("Query executed successfully")
+    except Error as e:
+        print(f"The error '{e}' occurred")
```
Same here:

```suggestion
    cursor.execute(query)
    connection.commit()
    print("Query executed successfully")
```

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+def execute_query_with_data(connection, query, data):
+    cursor = connection.cursor()
+
+    try:
```
and here and there are more below which I won't mention :D

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+    except Error as e:
+        print(f"The error '{e}' occurred")
+
+def insert_parsed(sql, list_of_lists):
+  vars = []
```
Lines below this as well:
Python cares a lot about white spaces. To change the amount of spaces used within a file can easily lead to unexpected behavior. I highly recommend using PEP8's recommendation of 4 spaces throughout this whole file.

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+#  - mediaPackageId, seriesId for oc_workflow
+print("Create tables...")
+create_workflow_table = f"""
+CREATE TABLE IF NOT EXISTS {workflow_table_name} (
```
Should there be the `IF NOT EXISTS`?

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+"""
+
+create_workflow_configuration_table = f"""
+CREATE TABLE IF NOT EXISTS {workflow_configuration_table_name} (
```
`IF NOT EXISTS`?

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+) ENGINE = InnoDB
+"""
+
+# ??? Missing "executionHistory" maybe?
```
?

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+create_workflow_configuration_table = f"""
+CREATE TABLE IF NOT EXISTS {workflow_configuration_table_name} (
+  workflow_id BIGINT(20),
+  key_part VARCHAR(255) NOT NULL,
+  value_part LONGTEXT,
```
Are there only parts of keys and values stored in here?

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+SELECT date_completed FROM oc_job WHERE operation="START_WORKFLOW"
+"""
+
+payloads = execute_read_query(connection, select_payload_from_job_table)
```
This will potentially load tens of thousands of workflows at the same time from the database. Not a good idea. Load them in chunks of maybe 100 workflows at a time.

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+SELECT date_created FROM oc_job WHERE operation="START_WORKFLOW"
+"""
+select_date_completed_from_job_table = """
+SELECT date_completed FROM oc_job WHERE operation="START_WORKFLOW"
```
Why not
```sql
select payload, date_created, date_completed from oc_job where operation="START_WORKFLOW"
```

lkiesow commented on docs/upgrade/11_to_12/workflows_mysql5.py:
```diff
@@ -0,0 +1,425 @@
+
+### Remove workflow XML from oc_job
+sql_update_job_payload_query = """
+UPDATE oc_job SET payload = %s where id = %s
```
Why not just as a single query:

```sql
update oc_job set payload = %s where operation="START_WORKFLOW"
```

lkiesow commented on docs/upgrade/11_to_12/workflows_postgresql.py:
```diff
@@ -0,0 +1,427 @@
+# Upgrade script for PostgreSQL databases
```
I'm leaving this out since the comments on this would be very much the same as the one on the MariaDB script

lkiesow reviewed (changes requested):
I'm not quite done with the review, but I'm though the migration part.

Arnei committed:
Merge remote-tracking branch 'upstream/develop' into workflows_from_xml_to_database_tables

Arnei committed:
Add release info

Arnei committed:
Fixed errors resulting from earlier merge

Arnei committed:
Forgot the remote module

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstance.java:
```diff
@@ -21,21 +21,162 @@
+  private String description;
+
+  @Column(name = "parent", nullable = true)
+  private Long parentId;
```
Assuming this corresponds to `oc_job.parent`, his should always be `null` since workflows are always root jobs.
Example from the ETH production system:

```sql
> select count(*) from oc_job where operation = 'START_WORKFLOW' and parent is not null;
+----------+
| count(*) |
+----------+
|        0 |
+----------+

> select count(*) from oc_job where operation = 'START_WORKFLOW' and parent is null;
+----------+
| count(*) |
+----------+
|   119037 |
+----------+
```

lkiesow commented on modules/index-service/src/main/java/org/opencastproject/index/service/impl/IndexServiceImpl.java:
```diff
@@ -760,13 +758,10 @@ private String startAddAssetWorkflow(JSONObject metadataJson, MediaPackage media
+      if (workflowService.mediaPackageHasActiveWorkflows(mpId)) {
+        logger.warn("Unable to start new workflow '{}' on archived media package '{}', existing workflow is running",
+                workflowDefId, mediaPackage);
+        throw new IllegalArgumentException("A workflow is already active for mp " + mpId + ", cannot start this workflow.");
```
No need for the warning if we then throw a similar error:

```suggestion
        throw new IllegalArgumentException("Unable to start new workflow '" + workflowDefId + "' on archived media package '" + mpId + "', existing workflow is running");
```

lkiesow commented on modules/index-service/src/main/java/org/opencastproject/index/service/impl/IndexServiceImpl.java:
```diff
@@ -1347,12 +1343,13 @@ public MetadataList updateEventMetadata(String id, MetadataList metadataList, El
+          Optional<WorkflowInstance> workflowInstance = workflowService.
+                  getRunningWorkflowInstanceByMediaPackage(event.getIdentifier(), Permissions.Action.WRITE.toString());
+          if (!workflowInstance.isPresent()) {
+            logger.error("No workflow instance for event {} found!", event.getIdentifier());
```
```suggestion
```

lkiesow commented on modules/index-service/src/main/java/org/opencastproject/index/service/impl/IndexServiceImpl.java:
```diff
@@ -1608,11 +1604,25 @@ private void updateWorkflowInstance(WorkflowInstance workflowInstance)
+                  getRunningWorkflowInstanceByMediaPackage(event.getIdentifier(), Permissions.Action.READ.toString());
+          if (!currentWorkflowInstance.isPresent()) {
+            logger.error("No workflow instance for event {} found!", event.getIdentifier());
+            throw new IndexServiceException("No workflow instance found for event " + event.getIdentifier());
```
```suggestion
            throw new IndexServiceException("No workflow instance found for event " + event.getIdentifier());
```

lkiesow commented on modules/index-service/src/main/java/org/opencastproject/index/service/impl/IndexServiceImpl.java:
```diff
@@ -1608,11 +1604,25 @@ private void updateWorkflowInstance(WorkflowInstance workflowInstance)
+        } catch (WorkflowDatabaseException e) {
+          logger.error("Unable to get current workflow instance for event with id {} from workflow service",
+                  event.getIdentifier(), e);
+          throw new IndexServiceException(e.getMessage(), e);
```
```suggestion
          throw new IndexServiceException("Unable to get current workflow instance for event with id " + event.getIdentifier() + " from workflow service", e);
```

lkiesow commented on modules/index-service/src/main/java/org/opencastproject/index/service/impl/IndexServiceImpl.java:
```diff
@@ -1608,11 +1604,25 @@ private void updateWorkflowInstance(WorkflowInstance workflowInstance)
+          throw new IndexServiceException(e.getMessage(), e);
+        } catch (UnauthorizedException e) {
+          logger.error("Not authorized to read mediapackage from workflow", event.getIdentifier(), e);
+          throw new IndexServiceException(e.getMessage(), e);
```
```suggestion
          throw new IndexServiceException("Not authorized to read mediapackage " + event.getIdentifier() + " from workflow", e);
```

lkiesow commented on modules/index-service/src/main/java/org/opencastproject/index/service/impl/IndexServiceImpl.java:
```diff
@@ -1608,11 +1604,25 @@ private void updateWorkflowInstance(WorkflowInstance workflowInstance)
+          throw new IndexServiceException(e.getMessage(), e);
+        } catch (WorkflowException e) {
+          logger.error("Unable to get event mediapackage {} from WorkflowService because", event.getIdentifier(), e);
+          throw new IndexServiceException(e.getMessage(), e);
```
```suggestion
          throw new IndexServiceException("Unable to get event mediapackage " + event.getIdentifier() + " from WorkflowService because", e);
```

lkiesow commented on modules/index-service/src/main/java/org/opencastproject/index/service/impl/IndexServiceImpl.java:
```diff
@@ -1979,8 +1966,9 @@ public void updateCommentCatalog(final Event event, final List<EventComment> com
+            Optional<WorkflowInstance> workflowInstance = workflowService.
+                    getRunningWorkflowInstanceByMediaPackage(event.getIdentifier(), Permissions.Action.WRITE.toString());
+            if (!workflowInstance.isPresent()) {
               logger.error("No workflow instance for event {} found!", event.getIdentifier());
```
```suggestion
```

lkiesow commented on modules/videoeditor-workflowoperation/src/main/java/org/opencastproject/workflow/handler/videoeditor/SilenceDetectionWorkflowOperationHandler.java:
```diff
@@ -182,6 +182,7 @@ public WorkflowOperationResult start(WorkflowInstance workflowInstance, JobConte
     Collection<Track> sourceTracks = trackSelector.select(mp, false);
     if (sourceTracks.isEmpty()) {
       logger.info("No source tracks found, skip silence detection");
+      workflowInstance.setMediaPackage(mp);
```
Why do we manually replace the media package here? We pass it on to `createResult()` which should do that for us automatically.

lkiesow replied:
Yep, this is set automatically:

https://github.com/opencast/opencast/blob/bf8a77cef5d9a25b37608b0e7f91f1579ad46655/modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java#L1606-L1609

lkiesow commented on modules/videoeditor-workflowoperation/src/main/java/org/opencastproject/workflow/handler/videoeditor/SilenceDetectionWorkflowOperationHandler.java:
```diff
@@ -262,6 +263,7 @@ public WorkflowOperationResult start(WorkflowInstance workflowInstance, JobConte
       }
     }
     logger.debug("Finished silence detection workflow operation for mediapackage {}", mp.getIdentifier().toString());
+    workflowInstance.setMediaPackage(mp);
```
Why do we manually replace the media package here? We pass it on to `createResult()` which should do that for us automatically.

lkiesow commented on modules/videoeditor-workflowoperation/src/main/java/org/opencastproject/workflow/handler/videoeditor/VideoEditorWorkflowOperationHandler.java:
```diff
@@ -341,6 +342,7 @@ public WorkflowOperationResult skip(WorkflowInstance workflowInstance, JobContex
     final boolean skipProcessing = BooleanUtils
             .toBoolean(worflowOperationInstance.getConfiguration(SKIP_PROCESSING_PROPERTY));
     if (skipProcessing) {
+      workflowInstance.setMediaPackage(mp);
```
Why do we manually replace the media package here? We pass it on to `createResult()` which should do that for us automatically.

lkiesow commented on modules/videoeditor-workflowoperation/src/main/java/org/opencastproject/workflow/handler/videoeditor/VideoEditorWorkflowOperationHandler.java:
```diff
@@ -376,6 +378,7 @@ public WorkflowOperationResult skip(WorkflowInstance workflowInstance, JobContex
       mp.addDerived(clonedTrack, sourceTrack);
     }
 
+    workflowInstance.setMediaPackage(mp);
```
Why do we manually replace the media package here? We pass it on to `createResult()` which should do that for us automatically.

lkiesow commented on modules/videoeditor-workflowoperation/src/main/java/org/opencastproject/workflow/handler/videoeditor/VideoEditorWorkflowOperationHandler.java:
```diff
@@ -472,6 +475,7 @@ public WorkflowOperationResult resume(WorkflowInstance workflowInstance, JobCont
     // to delivery format
     if (skipProcessing) {
       logger.info("VideoEdit workflow {} finished - smil file is {}", workflowInstance.getId(), smil.getId());
+      workflowInstance.setMediaPackage(mp);
```
Why do we manually replace the media package here? We pass it on to `createResult()` which should do that for us automatically.

lkiesow commented on modules/videoeditor-workflowoperation/src/main/java/org/opencastproject/workflow/handler/videoeditor/VideoEditorWorkflowOperationHandler.java:
```diff
@@ -591,6 +595,7 @@ public WorkflowOperationResult resume(WorkflowInstance workflowInstance, JobCont
     }
 
     logger.info("VideoEdit workflow {} finished", workflowInstance.getId());
+    workflowInstance.setMediaPackage(mp);
```
Why do we manually replace the media package here? We pass it on to `createResult()` which should do that for us automatically.

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/JaxbWorkflowInstance.java:
```diff
@@ -0,0 +1,174 @@
+  private String description;
+
+  @XmlElement(name = "parent", nillable = true)
+  private Long parentId;
```
I don't think we need this. See database definition for more details.

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/JaxbWorkflowInstance.java:
```diff
@@ -0,0 +1,174 @@
+  protected String mediaPackageId;
+
+  @XmlElement
+  protected String seriesId;
```
What do we need this for?

lkiesow replied:
We don't need that except maybe for #3377?

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/JaxbWorkflowInstance.java:
```diff
@@ -0,0 +1,174 @@
+    this.dateCreated = workflow.getDateCreated();
+    this.dateCompleted = workflow.getDateCompleted();
+    this.mediaPackage = workflow.getMediaPackage();
+    this.operations = Optional.ofNullable(workflow.getOperations())
```
I would expect `workflow.getOperations(` to return an empty list by itself and not `null`.

lkiesow replied:
And it is guaranteed that `getOperations()` does not return `null`

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/JaxbWorkflowInstance.java:
```diff
@@ -0,0 +1,174 @@
+            .map(operation -> new JaxbWorkflowOperationInstance(operation))
+            .collect(Collectors.toList());
+    if (workflow.getConfigurations() != null) {
+      this.configurations = Optional.ofNullable(workflow.getConfigurations().entrySet())
```
If `getConfigurations` returns a `Map`, `entrySet` should never be `null`.

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/JaxbWorkflowOperationInstance.java:
```diff
@@ -0,0 +1,226 @@
+    this.description = operation.getDescription();
+    if (operation.getConfigurations() != null) {
+      this.configurations = Optional.ofNullable(operation.getConfigurations().entrySet())
+              .orElseGet(Collections::emptySet)
```
`entrySet` should never return `null`:

```suggestion
      this.configurations = operation.getConfigurations().entrySet()
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstance.java:
```diff
@@ -21,21 +21,162 @@
+  private Date dateCompleted = null;
+
+  @Lob
+  @Basic(fetch = FetchType.LAZY)
```
Does this have any effect on a native field? I only know this from collections which need to be fetched from other tables. And if this has an effect, then why not fetch this?

lkiesow replied:
My assumption seems to be correct. It only works on collections. Dropping this here.

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstance.java:
```diff
@@ -21,21 +21,162 @@
+          name = "oc_workflow_configuration",
+          joinColumns = @JoinColumn(name = "workflow_id")
+  )
+  @MapKeyColumn(name = "key_part", nullable = false)
```
```suggestion
  @MapKeyColumn(name = "configuration_key", nullable = false)
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstance.java:
```diff
@@ -21,21 +21,162 @@
+  )
+  @MapKeyColumn(name = "key_part", nullable = false)
+  @Lob
+  @Column(name = "value_part")
```
```suggestion
  @Column(name = "configuration_value")
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstance.java:
```diff
@@ -21,21 +21,162 @@
 
-  enum WorkflowState {
+  @Transient
+  protected boolean initialized = false;
```
Looks weird. What is this used for? Need to investigate.

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstance.java:
```diff
@@ -63,135 +204,400 @@ public WorkflowState unmarshal(String val) {
+    } catch (MediaPackageException e) {
+      // TODO: Error handling
+      logger.error("Error: ", e);
+    }
```
Why not throw this exception?

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstance.java:
```diff
@@ -63,135 +204,400 @@ public WorkflowState unmarshal(String val) {
+    if (operations == null || operations.isEmpty())
+      return;
+
+    // Operation's position, so we fix it here
```
Unfinished sentence?

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstance.java:
```diff
@@ -63,135 +204,400 @@ public WorkflowState unmarshal(String val) {
+
+        WorkflowOperationInstance operation = operations.get(position);
+
+        switch (operation.getState()) {
```
This is a very complicated way of finding the first active operation. Why not just something like:

```java
for (var op: operations) {
  if (op.isActive()) {
    return op;
  }
}
return null;
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstance.java:
```diff
@@ -63,135 +204,400 @@ public WorkflowState unmarshal(String val) {
    */
-  WorkflowOperationInstance getCurrentOperation() throws IllegalStateException;
+  public Set<String> getConfigurationKeys() {
+    Set<String> keys = new TreeSet<String>();
```
Why not just return `configurations.keySet()`?

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstance.java:
```diff
@@ -63,135 +204,400 @@ public WorkflowState unmarshal(String val) {
-   * The {@link MediaPackage} being worked on by this workflow instance.
-   */
-  MediaPackage getMediaPackage();
+  private void addConfiguration(String key, String value) {
```
Isn't this basically the same as `setConfiguration` above, just without the check?

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstance.java:
```diff
@@ -63,135 +204,400 @@ public WorkflowState unmarshal(String val) {
+  public void extend(WorkflowDefinition workflowDefinition) {
+    if (!workflowDefinition.getOperations().isEmpty()) {
+      for (WorkflowOperationDefinition entry : workflowDefinition.getOperations()) {
+        operations.add(new WorkflowOperationInstance(entry, -1));
```
Why `-1`?

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstance.java:
```diff
@@ -63,135 +204,400 @@ public WorkflowState unmarshal(String val) {
+        offset++;
+        operations.add(after.getPosition() + offset, new WorkflowOperationInstance(entry, -1));
+      }
+      setOperations(operations);
```
Didn't we already added them to the class member variable `operations`?

lkiesow replied:
Was used to set the operation's workflow instance

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowOperationInstance.java:
```diff
@@ -22,97 +22,452 @@
+          name = "oc_workflow_operation_configuration",
+          joinColumns = @JoinColumn(name = "workflow_operation_id")
+  )
+  @MapKeyColumn(name = "key_part", nullable = false)
```
```suggestion
  @MapKeyColumn(name = "configuration_key", nullable = false)
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowOperationInstance.java:
```diff
@@ -22,97 +22,452 @@
+  )
+  @MapKeyColumn(name = "key_part", nullable = false)
+  @Lob
+  @Column(name = "value_part")
```
```suggestion
  @Column(name = "configuration_value")
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowOperationInstance.java:
```diff
@@ -22,97 +22,452 @@
+  protected Map<String, String> configurations;
+
+  @Column(name = "holdurl")
+  protected String holdStateUserInterfaceUrl;
```
Can we drop this? I don't think it's actually used any longer.

lkiesow replied:
Seems like we can

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowOperationInstance.java:
```diff
@@ -22,97 +22,452 @@
+    } else if (OperationState.FAILED.equals(state)) {
+      this.dateCompleted = now;
+    } else if (OperationState.SUCCEEDED.equals(state)) {
+      this.dateCompleted = now;
```
collapse

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowOperationInstance.java:
```diff
@@ -22,97 +22,452 @@
+   */
+  @Override
+  public Set<String> getConfigurationKeys() {
+    Set<String> keys = new TreeSet<String>();
```
Why do we copy these?

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowOperationInstance.java:
```diff
@@ -22,97 +22,452 @@
    * when exceptions are thrown during an operation.
    */
-  boolean isFailWorkflowOnException();
+  public boolean isFailWorkflowOnException() {
+    return failWorkflowOnException;
+  }
```
Is this ever false?

lkiesow replied:
It is. This corresponds to `fail-on-error`. Let's make sure we just use one name for this to avoid confusion.

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowService.java:
```diff
@@ -342,4 +343,24 @@ void cleanupWorkflowInstances(int lifetime, WorkflowInstance.WorkflowState state
+
+  boolean mediaPackageHasActiveWorkflows(String mediaPackageId) throws WorkflowDatabaseException;
+
+  List<WorkflowInstance> getWorkflowInstancesByMediaPackage(String mediaPackageId) throws WorkflowDatabaseException;
```
JavaDoc

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowServiceDatabase.java:
```diff
@@ -0,0 +1,125 @@
+   * @throws WorkflowDatabaseException
+   *           if there is a problem communicating with the underlying data store
+   */
+  List<WorkflowInstance> getAllWorkflowInstancesOrganizationIndependent() throws WorkflowDatabaseException;
```
If this is used for the index rebuild, we probably do not want to get all workflows at once. There are potentially thousands  of workflows in a system.

lkiesow replied:
Seems like it's used for exactly that, which brings us back to the problem fixed by https://github.com/opencast/opencast/pull/104 which still gets all the workflows, but only as a list of strings and then deserializes them if they are needed. This will get all workflows and build Java objects from all of them at once. Not a good idea.

lkiesow replied:
Digging deeper, this method is used only for the workflow Solr index. Given that that one is to be removed, let's ignore this for now. I added a warning to the logs though to be sure we don't forget.

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowServiceDatabase.java:
```diff
@@ -0,0 +1,125 @@
+   * @throws WorkflowDatabaseException
+   *           if there is a problem communicating with the underlying data store
+   */
+  List<WorkflowInstance> getAllWorkflowInstances(int limit, int offset) throws WorkflowDatabaseException;
```
Same as `getAllWorkflowInstancesOrganizationIndependent`

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowServiceDatabaseImpl.java:
```diff
@@ -0,0 +1,322 @@
+      em = emf.createEntityManager();
+
+      String orgId = securityService.getOrganization().getId();
+      Query q = em.createNamedQuery("Workflow.workflowById");
```
`TypedQuery`?

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowServiceDatabaseImpl.java:
```diff
@@ -0,0 +1,322 @@
+    try {
+      em = emf.createEntityManager();
+
+      Query query = em.createNamedQuery("Workflow.findAllOrganizationIndependent");
```
`TypedQuery`

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowServiceDatabaseImpl.java:
```diff
@@ -0,0 +1,322 @@
+      Query query = em.createNamedQuery("Workflow.findAllOrganizationIndependent");
+
+      List<WorkflowInstance> workflowInstances = query.getResultList();
+      return workflowInstances;
```
```suggestion
      return query.getResultList();
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowServiceDatabaseImpl.java:
```diff
@@ -0,0 +1,322 @@
+    try {
+      em = emf.createEntityManager();
+
+      Query query = em.createNamedQuery("Workflow.findAll", String.class);
```
`TypedQuery` → _more down below._

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowServiceDatabaseImpl.java:
```diff
@@ -0,0 +1,322 @@
+      query.setParameter("mediaPackageId", mediaPackageId);
+
+      List<WorkflowInstance> workflowInstances = query.getResultList();
+      return workflowInstances;
```
```suggestion
      return query.getResultList();
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowServiceDatabaseImpl.java:
```diff
@@ -0,0 +1,322 @@
+      if (tx.isActive()) {
+        tx.rollback();
+      }
+      throw new WorkflowDatabaseException(e);
```
```suggestion
      if (tx.isActive()) {
        tx.rollback();
      }
      throw new WorkflowDatabaseException("Could not update workflow with ID '" + instance.getId() + "'", e);
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowServiceDatabaseImpl.java:
```diff
@@ -0,0 +1,322 @@
+      if (tx.isActive()) {
+        tx.rollback();
+      }
+      throw new WorkflowDatabaseException(e);
```
```suggestion
      if (tx.isActive()) {
        tx.rollback();
      }
      throw new WorkflowDatabaseException("Could not delete workflow with ID '" + instance.getId() + "'", e);
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowSetImpl.java:
```diff
@@ -65,15 +66,19 @@ public class WorkflowSetImpl implements WorkflowSet {
   }
 
+  public WorkflowSetImpl(List<WorkflowInstance> workflows) {
+    this.resultSet = workflows.stream().map(workflow -> new JaxbWorkflowInstance(workflow)).collect(Collectors.toList());
```
```suggestion
    this.resultSet = workflows.stream().map(JaxbWorkflowInstance::new).collect(Collectors.toList());
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowSetImpl.java:
```diff
@@ -65,15 +66,19 @@ public class WorkflowSetImpl implements WorkflowSet {
-    return resultSet == null || resultSet.size() == 0 ? new WorkflowInstance[0] : resultSet
-            .toArray(new WorkflowInstance[resultSet.size()]);
+  public List<WorkflowInstance> getItems() {
+    return resultSet == null || resultSet.size() == 0 ? new ArrayList<>() : resultSet.stream().map(workflow -> workflow.toWorkflowInstance()).collect(Collectors.toList());
```
_simplify…_

Alaschinger replied:
Do it this way?
```suggestion
List<WorkflowInstance> items = new ArrayList<WorkflowInstance>();
if (resultSet != null && resultSet.size() != 0)
  items = resultSet.stream().map(workflow -> workflow.toWorkflowInstance()).collect(Collectors.toList());
return items;
```

lkiesow replied:
I used something similar:

```java
if (resultSet == null) {                                                                                                                                                                                                                 
  return Collections.emptyList();                                                                                                                                                                                                        
}                                                                                                                                                                                                                                        
return resultSet.stream()                                                                                                                                                                                                                
    .map(JaxbWorkflowInstance::toWorkflowInstance)                                                                                                                                                                                       
    .collect(Collectors.toList());
```

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/endpoint/WorkflowRestService.java:
```diff
@@ -306,6 +309,70 @@ public Response getConfigurationPanel(@QueryParam("definitionId") String definit
+      throw new WebApplicationException(Status.INTERNAL_SERVER_ERROR);
+    } catch (UnauthorizedException e) {
+      throw new WebApplicationException(Status.INTERNAL_SERVER_ERROR);
+    } catch (WorkflowException e) {
```
```suggestion
    } catch (WorkflowDatabaseException | UnauthorizedException | WorkflowException e) {
```

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -669,15 +657,34 @@ protected WorkflowInstance updateConfiguration(WorkflowInstance instance, Map<St
+      for (String configKey : instance.getConfigurationKeys()) {
+              String configValue = WorkflowConditionInterpreter.replaceVariables(instance.getConfiguration(configKey),
+                      systemVariableGetter, wfProperties, false);
+              instance.setConfiguration(configKey, configValue);
```
indentation

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -669,15 +657,34 @@ protected WorkflowInstance updateConfiguration(WorkflowInstance instance, Map<St
+        }
+      }
+
+      // Previous approach
+//      String xml = WorkflowConditionInterpreter.replaceVariables(XmlWorkflowParser.toXml(instance),
+//              systemVariableGetter, wfProperties, false);
+//      return WorkflowParser.parseWorkflowInstance(xml);
```
```suggestion
```

lkiesow replied:
Conditions are basically broken → need to fundamentally change how this works

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -992,80 +999,64 @@ public void remove(long workflowInstanceId) throws WorkflowDatabaseException, No
       WorkflowInstance.WorkflowState state = instance.getState();
-      if (state != WorkflowState.SUCCEEDED && state != WorkflowState.FAILED && state != WorkflowState.STOPPED) {
+      if (state != WorkflowState.SUCCEEDED && state != WorkflowState.FAILED && state != WorkflowState.STOPPED)
         if (!force) {
```
Control structure weirdness

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -992,80 +999,64 @@ public void remove(long workflowInstanceId) throws WorkflowDatabaseException, No
+        // Second, remove jobs related to a operation which belongs to the workflow instance
+        List<WorkflowOperationInstance> operations = instance.getOperations();
+        List<Long> jobsToDelete = new ArrayList<>();
+        for (WorkflowOperationInstance op : operations) {
```
`CASCADE`?

lkiesow replied:
Cascade would be good but this would need cascading in the service registry database table → a problem for another time

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -1357,15 +1345,15 @@ public void update(final WorkflowInstance workflowInstance) throws WorkflowExcep
         throw new WorkflowDatabaseException(e);
       } catch (NotFoundException e) {
         logger.error("Job for workflow %s not found in service registry", workflowInstance.getId());
         throw new WorkflowDatabaseException(e);
```
```suggestion
        throw new WorkflowDatabaseException("Job for workflow " + workflowInstance.getId() + " not found in service registry", e);
```

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -1357,15 +1345,15 @@ public void update(final WorkflowInstance workflowInstance) throws WorkflowExcep
-                "Update of workflow job %s in the service registry failed, service registry and workflow index may be out of sync",
+                "Update of workflow job %s in the service registry failed, service registry and workflow table may be out of sync",
                 workflowInstance.getId());
         throw new WorkflowDatabaseException(e);
```
```suggestion
        throw new WorkflowDatabaseException("Update of workflow job " + workflowInstance.getId() + " in the service registry failed, service registry and workflow table may be out of sync", e);
```

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -1357,15 +1345,15 @@ public void update(final WorkflowInstance workflowInstance) throws WorkflowExcep
-                "Update of workflow job %s in the service registry failed, service registry and workflow index may be out of sync",
+                "Update of workflow job %s in the service registry failed, service registry and workflow table may be out of sync",
                 job.getId());
         throw new WorkflowException(e);
```
logger/error thingy

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -1737,28 +1755,40 @@ public boolean isReadyToAccept(Job job) throws UndispatchableJobException {
+    } catch (UnauthorizedException e) {
+      logger.error("Authorization denied while requesting to loading workflow instance %s: %s", job.getId(),
+              e.getMessage());
+      throw new UndispatchableJobException(e);
```
```suggestion
      throw new UndispatchableJobException("Authorization denied while requesting to loading workflow instance " + job.getId(), e);
```

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -1737,28 +1755,40 @@ public boolean isReadyToAccept(Job job) throws UndispatchableJobException {
-      logger.error("Error loading workflow instance %s: %s", job.getId(), e.getMessage());
+      logger.error("An database error occured while checking if a workflow is already active %s: %s", job.getId(),
+              e.getMessage());
+      throw new UndispatchableJobException(e);
```
```suggestion
      throw new UndispatchableJobException("An database error occured while checking if a workflow is already active " + job.getId(), e);
```

lkiesow commented on modules/workflow-service-remote/src/main/java/org/opencastproject/workflow/remote/WorkflowServiceRemoteImpl.java:
```diff
@@ -277,6 +278,73 @@ public WorkflowSet getWorkflowInstancesForAdministrativeRead(WorkflowQuery q) th
+          throws WorkflowDatabaseException {
+    List<NameValuePair> queryStringParams = new ArrayList<>();
+
+    queryStringParams.add(new BasicNameValuePair("mediaPackageId", mediaPackageId));
+
+    HttpGet get = new HttpGet("/mediaPackage/" + mediaPackageId + "/instances.xml");
```
Why does this request need the media package identifier twice?

lkiesow commented on modules/workflow-service-remote/src/main/java/org/opencastproject/workflow/remote/WorkflowServiceRemoteImpl.java:
```diff
@@ -277,6 +278,73 @@ public WorkflowSet getWorkflowInstancesForAdministrativeRead(WorkflowQuery q) th
+          throws WorkflowException, UnauthorizedException, WorkflowDatabaseException {
+    List<NameValuePair> queryStringParams = new ArrayList<>();
+
+    queryStringParams.add(new BasicNameValuePair("mediaPackageId", mediaPackageId));
```
same here

Alaschinger replied:
@lkiesow Maybe I overlooked something, but you didn't mean the string "mediaPackageId" followed by the actual mediaPackageId, did you?

lkiesow replied:
I meant exactly that, why?

lkiesow commented on modules/workflow-service-remote/src/main/java/org/opencastproject/workflow/remote/WorkflowServiceRemoteImpl.java:
```diff
@@ -277,6 +278,73 @@ public WorkflowSet getWorkflowInstancesForAdministrativeRead(WorkflowQuery q) th
+  public boolean mediaPackageHasActiveWorkflows(String mediaPackageId) throws WorkflowDatabaseException {
+    List<NameValuePair> queryStringParams = new ArrayList<>();
+
+    queryStringParams.add(new BasicNameValuePair("mediaPackageId", mediaPackageId));
```
and here

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/AddCatalogWorkflowOperationHandler.java:
```diff
@@ -151,6 +151,8 @@ else if (collBehavior == CatalogTypeCollisionBehavior.SKIP) {
       mpe.addTag(tag);
     }
 
+    wInst.setMediaPackage(mp);
```
Shouldn't `createResult` do this?

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/CloneWorkflowOperationHandler.java:
```diff
@@ -117,6 +117,7 @@ public WorkflowOperationResult start(WorkflowInstance workflowInstance, JobConte
     // Make sure either one of tags or flavors are provided
     if (sourceTagsOption.isEmpty() && sourceFlavorOptionList.isEmpty()) {
       logger.info("No source tags or flavors have been specified, not matching anything. Operation will be skipped.");
+      workflowInstance.setMediaPackage(mediaPackage);
```
`createResult` should do this

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/CloneWorkflowOperationHandler.java:
```diff
@@ -151,6 +152,7 @@ public WorkflowOperationResult start(WorkflowInstance workflowInstance, JobConte
     if (elements.size() == 0) {
       // If no one found, we skip the operation
       logger.debug("No matching elements found, skipping operation.");
+      workflowInstance.setMediaPackage(mediaPackage);
```
`createResult` should do this

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/CloneWorkflowOperationHandler.java:
```diff
@@ -168,6 +170,7 @@ public WorkflowOperationResult start(WorkflowInstance workflowInstance, JobConte
       }
     }
 
+    workflowInstance.setMediaPackage(mediaPackage);
```
`createResult` should do this

lkiesow committed:
Remove ParentId From Workflow Table

Workflow jobs do not have parents so this will always be `null` anyway.

Example from the ETH production system:

```sql
> select count(*) from oc_job where operation = 'START_WORKFLOW' and parent is not null;
+----------+
| count(*) |
+----------+
|        0 |
+----------+

> select count(*) from oc_job where operation = 'START_WORKFLOW' and parent is null;
+----------+
| count(*) |
+----------+
|   119037 |
+----------+
```

lkiesow committed:
Logging and Errors

This patches up a number of error messages, making sure that thrown
exceptions have proper error messages and they are not separated from
the actual error.

lkiesow committed:
Don't set media package twice

The workflow service will already update the workflow instance with the
resulting media package. No need to do this twice.

lkiesow committed:
Fix Workflow Return Values

Some workflow operations seem to not return updated media packages,
instead relying on side effects of the modification of existing media
packages.

lkiesow committed:
…

lkiesow committed:
Minor fixes

lkiesow committed:
No Hold Action Metadata

They are not used any more.

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowOperationInstance.java:
```diff
@@ -21,98 +21,394 @@
+  @Column(name = "value_part")
+  protected Map<String, String> configurations;
+
+  @Column(name = "failOnError")
+  protected boolean failWorkflowOnException;
+
+  @Column(name = "if_condition")
+  protected String executeCondition;
```
Mixing camel case (which usually is not carried to the databases) and snake case.
I suggest using the latter for everything.

lkiesow committed:
…

lkiesow committed:
currentWorkflowOp

lkiesow committed:
…

lkiesow committed:
…

lkiesow committed:
names

lkiesow committed:
database column names

lkiesow committed:
typed-queries

lkiesow committed:
…

lkiesow committed:
Remove skipCondition

lkiesow committed:
simplify

lkiesow committed:
unless-tests

lkiesow committed:
execution condition

lkiesow committed:
Remove debug error logs

lkiesow committed:
exceptions

lkiesow committed:
exception handling

lkiesow committed:
remote-cleanup

lkiesow committed:
more exception handling

lkiesow committed:
even more exception handling

lkiesow committed:
Fix cover-image workflow configuration

lkiesow commented on docs/upgrade/11_to_12/workflows_postgresql.py:
```diff
@@ -0,0 +1,427 @@
+
+### Remove workflow XML from oc_job
+sql_update_job_payload_query = """
+UPDATE oc_job SET payload = %s where id = %s
+"""
+for id in ids:
+  execute_query_with_data(connection, sql_update_job_payload_query, (id, id))
```
Why set `payload` to the entities primary key? If it's identical, we could just set it to `null`.

lkiesow committed:
MariaDB Migration

lkiesow committed:
Finish MariaDB Migration

lkiesow committed:
Python 3.6 compatibility

lkiesow commented on docs/upgrade/11_to_12/workflows_postgresql.py:
```diff
@@ -0,0 +1,427 @@
+#  - mediaPackageId, seriesId for oc_workflow
+print("Create tables...")
+create_workflow_table = f"""
+CREATE TABLE IF NOT EXISTS {workflow_table_name} (
```
Foreign keys seem to be missing here.

lkiesow committed:
Finished migration scripts

lkiesow committed:
Merge develop

lkiesow committed:
Removed comment

lkiesow committed:
Use MariaDB instead of MySQL

lkiesow commented:
I ran several migration tests on this with both PostgreSQL and MariaDB. It now works as expected and there should be no problem, even with larger databases. A database with 120k workflows should take 20-30min to migrate. That's okay. Memory consumption is negligible (< 100 MB) for the migration script.

lkiesow committed:
Use workflow job id as workflow id

Arnei commented on docs/upgrade/11_to_12/workflows_mariadb.py:
```diff
@@ -0,0 +1,379 @@
+  `workflow_id` bigint(20) DEFAULT NULL,
+  `configuration_value` longtext DEFAULT NULL,
+  `configuration_key` varchar(255) DEFAULT NULL,
+  KEY `FK_oc_workflow_configuration_workflow_id` (`workflow_id`),
+  CONSTRAINT `FK_oc_workflow_configuration_workflow_id` FOREIGN KEY (`workflow_id`) REFERENCES `oc_workflow` (`id`)
```
Why do we need to manually describe index keys? Is `INDEX` not reliable?

Arnei commented on docs/upgrade/11_to_12/workflows_mariadb.py:
```diff
@@ -0,0 +1,379 @@
+user = "opencast"
+password = "dbpassword"
+host = "127.0.0.1"
+database = "opencast"
+
+# Constants
```
Per your earlier suggestion, do we want to add "# DO NOT EDIT ANYTHING BELOW"?

lkiesow replied:
Good point, we could. I add it to the todo. We wanted to switch to the operation job IDs anyway.

Arnei commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowOperationInstance.java:
```diff
@@ -21,98 +21,380 @@
+  protected RetryStrategy retryStrategy;
+
+  @ManyToOne(fetch = FetchType.LAZY)
+  @JoinColumn(name = "workflow_id", nullable = false)
```
Is the `@JoinColumn` here necessary? The relation should already be identified by the corresponding `mappedBy` in WorkflowInstance.java, right?

lkiesow replied:
It's not really necessary. This is just to get a nice column name like `workflow_id`

Arnei reviewed (commented):
Thanks Lars for bringing this PR up to mergeable speed! The comments below are just some things I noticed while reviewing my blunders.
